torch/cuda/__init__.py

```diff
@@ -317,41 +318,93 @@ def can_device_access_peer(device: _device_t, peer_device: _device_t) -> bool:
-        with device(stream.device):
-            dst_prev_stream = current_stream()
+    def __enter__(self):
+        self.idx = _get_device_index(device=None, optional=True)
+        # If there is no CUDA device available, return
+        if self.idx == -1:
+            return
 
-    torch._C._cuda_setStream(stream._cdata)
-    try:
-        yield
-    finally:
-        if src_prev_stream.device != stream.device:
-            torch._C._cuda_setStream(dst_prev_stream._cdata)
-        torch._C._cuda_setStream(src_prev_stream._cdata)
+        # Local cur_stream variable for type refinement
+        cur_stream = self.stream
+        # Return if stream is None
+        if cur_stream is None:
+            return
+        self.src_prev_stream = torch.cuda.current_stream(self.idx)
+
+        # Eager mode and script mode APIs to access the stream device and the device index
+        # vary. Hence, we have to check the mode we are in and call the respective API.
+        #
+        # Note: The API's don't differ in functionality, only differ at the API level.
+        # In Eager Mode, the stream object doesn't have a function to access device index,
+        # hence we use `_get_device_index`. In script mode, device_index is easily
+        # accessible by calling stream.device_index(). With device, in Eager mode to access
+        # device of the stream can be accessed by stream.device while in JIT we need to use
+        # stream.device(). But both return the device object of type `torch.device`
+        if torch.jit.is_scripting():
+            self.prev_stream_device_index = self.src_prev_stream.device_index()  # type: ignore
+            self.cur_stream_device_index = cur_stream.device_index()
+            cur_stream_device = cur_stream.device()
+        else:
+            self.prev_stream_device_index = _get_device_index(self.src_prev_stream.device, optional=True)
+            self.cur_stream_device_index = _get_device_index(cur_stream.device, optional=True)
+            cur_stream_device = cur_stream.device
+
+        # If the stream is not on the current device, then change the device
+        # and set the current stream on the device
+        if self.prev_stream_device_index != self.cur_stream_device_index:
+            with device(cur_stream_device):
+                self.dst_prev_stream = torch.cuda.current_stream(self.cur_stream_device_index)
+        torch.cuda.set_stream(cur_stream)  # type: ignore
+
+    def __exit__(self, type: Any, value: Any, traceback: Any):
+        # Local cur_stream variable for type refinement
+        cur_stream = self.stream
+        # If stream is None or no CUDA device available, return
+        if cur_stream is None or self.idx == -1:
+            return
 
+        # If the stream was not on the current device, restore the previous stream on
+        # the destination device and also reset the current device to the previous device.
+        # Set the current stream on the device to the src_prev_stream
```

Comment:
This comment appears to be out of date now that `set_stream` has been updated to set device if needed.